examples/darts/uno/default_model.txt

```diff
@@ -0,0 +1,18 @@
+[Global_Params]
+model_name = 'darts_uno'
+unrolled = False 
+data_url = 'ftp.mcs.anl.gov/pub/candle/public/benchmarks/Pilot1/uno/' 
+savepath = '.'
+log_interval = 10
+train_data = 'top_21_auc_1fold.uno.h5'
```

Comment:
Harry shared this data file with me, but I do not know how to best get a hold of that data for our example. I have a way of getting around it, but it makes everyone download the data into the example directory. I would like to get this fixed before merging.

Reply:
If I understand the question, the issue looks to be simply that the data_url has to be a valid url, for example in P1B1 it is (note the http://). 
data_url = 'http://ftp.mcs.anl.gov/pub/candle/public/benchmarks/P1B1/'
Is that what you are needing? I looked at Uno and it looks like they are doing a bunch of nonstandard things (I will have to try to fix it). 

If the file you are using in one that is used by Uno anyway, then the cache_subdir (last argument) should just be 'Pilot1' to avoid making an extra copy (It will check for the existence of the local copy and not download more than once)

so 
     fpath = candle.fetch_file(path + gParameters['train_data'], 'Pilot1')

Reply:
Ah, you are right. I goofed up and left off the `http://`. Rerunning that example with that data, it downloads that file now. Though it looks like it is going to take ~10 minutes to download it.

Reply:
Let me try with that cache dir.

Reply:
For whatever reason, it doesn't seem like Uno is using the candle file functions in a standard way, so it seems to be a bad example from that standpoint. It also uses huge data sets, so not the best example for any sort of quick demo.

Reply:
Setting the `subdir="Pilot1` in `candle.fetch_file`:

```python
fpath = candle.fetch_file(
    path + gParameters['train_data'], subdir='Pilot1'
)
```
then it starts to download that file. Does that look good to you?

Reply:
That seems right. We flattened the Data directory organization to make it easier to share between benchmarks (so anything from Pilot1 would use data in /Data/Pilot1/). If Uno was using the calls properly this would allow them to see this file if it had been used previously.

Reply:
Hmm... problem is that if `candle.fetch_file` doesn't seem to create the necessary directories for you with those arguments. If a user were to run this example on their own machine, it will fail:
```
Downloading data from http://ftp.mcs.anl.gov/pub/candle/public/benchmarks/Pilot1/uno/top_21_auc_1fold.uno.h5
6818488320/6818556817 [============================>.] - ETA: 0s
Traceback (most recent call last):
  File "uno_example.py", line 200, in <module>
    main()
  File "uno_example.py", line 196, in main
    run(params)
  File "uno_example.py", line 53, in run
    train_data = darts.Uno(datapath, 'train', download=True)
  File "/Users/ygx/src/ecp/Benchmarks/common/darts/datasets/uno.py", line 47, in __init__
    self.download()
  File "/Users/ygx/src/ecp/Benchmarks/common/darts/datasets/uno.py", line 157, in download
    makedir_exist_ok(self.raw_folder)
  File "/Users/ygx/src/ecp/Benchmarks/common/darts/datasets/utils.py", line 39, in makedir_exist_ok
    os.makedirs(dirpath)
  File "/Users/ygx/.local/opt/anaconda3/envs/darts/lib/python3.7/os.py", line 213, in makedirs
    makedirs(head, exist_ok=exist_ok)
  File "/Users/ygx/.local/opt/anaconda3/envs/darts/lib/python3.7/os.py", line 223, in makedirs
    mkdir(name, mode)
NotADirectoryError: [Errno 20] Not a directory: '/Users/ygx/src/ecp/Benchmarks/common/../Data/Pilot1/top_21_auc_1fold.uno.h5/Uno'
```

Reply:
In the designed use, the url should contain all the directory info and the file should just be the filename itself (with no subdirectories).

Reply:
There is a get_file function that is called by fetch_file, we might need to take a look at that but it should deal with this correctly.

Reply:
Does it matter that I am calling this from `Candle/examples/darts/uno` and not a Benchmarks directory?

Reply:
There might be an issue with it being 3 levels down from Benchmarks instead of 2, but that path is correct until Pilot1. The problem is that it shouldn't be trying to generate/access those subdirectories (it should put the file in /Data/Pilot1/). What is in that directory?

Reply:
I am calling this function from `Benchmarks/examples/darts/uno`. Here is the directory that I am calling that function from. This is showing the directory structure if I have the data download directly into that location:
```
  .
├──   __pycache__
│  └──   example_setup.cpython-37.pyc
├──   default_model.txt
├──   example_setup.py
├──   ftp.mcs.anl.gov
│  └──   pub
│     └──   candle
│        └──   public
│           └──   benchmarks
│              └──   Pilot1
│                 └──   uno
│                    └──   top_21_auc_1fold.uno.h5
│                       └──   Uno
│                          ├──   processed
│                          │  ├──   test_data.pt
│                          │  ├──   test_labels.pt
│                          │  ├──   train_data.pt
│                          │  └──   train_labels.pt
│                          └──   raw
│                             └──   top_21_auc_1fold.uno.h5
├──   Output
│  └──   EXP000
│     └──   RUN000
├──   README.rst
└──   uno_example.py
```

Reply:
That path `ft.pmcs.anl.gov/...` is added by a data class I have in DARTS that will download the data. If I use the `candle.fetch_data` it fails when calling from this directory.

Reply:
I was asking what is in the local Data directory. I think I didn't realize that the h5 file is actually a directory? If that is the case then it the fetch won't work (you can look at the function and it has switches for untarring a tarfile). In that case the way to do it is to tar up the top_21_auc_1fold.uno.h5 directory and fetch_file will untar it to replicate the directory structure.

Reply:
I think there are 2 separate issues. The lack of flexibility of local directory structure (this is because of the way we deal with python path, could be fixed by setting a DATA_LOCAL_DIR variable of some sort, similarly for CANDLE_LIB_DIR? And the issue of downloading multiple files.

Reply:
In fact I did something similar in examples/M16_test.py. This requires a bunch of files from various places and I replicated the directory structure locally. Like this;

# download all the gene_set files needed
data_url = 'http://ftp.mcs.anl.gov/pub/candle/public/benchmarks/Pilot1/uno/Candle_Milestone_16_Version_12_15_2019/Data/Gene_Sets/MSigDB.v7.0/'
for gene_set_category in ['c2.cgp','c2.cp.biocarta','c2.cp.kegg','c2.cp.pid','c2.cp.reactome','c5.bp','c5.cc','c5.mf','c6.all']:
    for gene_name_type in ['entrez', 'symbols']:
        file_name = gene_set_category+'.v7.0.'+gene_name_type+'.gmt'
        local_file = candle.get_file(file_name, data_url+file_name, cache_subdir='examples/Gene_Sets/MSigDB.v7.0')
# extract base directory for gene_set data files
data_dir = local_file.split(file_name)[0]
print('Gene Set data is locally stored at ', data_dir)

There is also an option Andrew wanted in get_file that allows you to specify data_dir as a complete path, so that might be what you need (they use a different directory structure on BioWolf). Making a local data directory is probably fine for now, it might just mean files get replicated in /Data.

Reply:
Cool, cool. I think this would be the simplest solution on my end: https://github.com/ECP-CANDLE/Benchmarks/pull/61/commits/736cb65a204bdb453fed1c032609b33d0a5ed410

How does that sound?

Reply:
If 'this' is using a local directory examples/darts/uno/data then I am fine with it. I still don't think it should fail since the get_file path is relative to common and you should be able to set that relative path with your 'include candle'

if datadir is None:
        file_path = os.path.dirname(os.path.realpath(__file__))
        datadir_base = os.path.expanduser(os.path.join(file_path, '..', 'Data'))
        datadir = os.path.join(datadir_base, cache_subdir)

    if not os.path.exists(datadir):
        os.makedirs(datadir)